File: Sprint-1/Python/remove_duplicates/remove_duplicates.py
from typing import List, Sequence, TypeVar

ItemType = TypeVar("ItemType")


def remove_duplicates(values: Sequence[ItemType]) -> List[ItemType]:
    """
    Remove duplicate values from a sequence, preserving the order of the first occurrence of each value.

    Time complexity:
    Space complexity:
    Optimal time complexity:
    """
    seen = set()
    unique_items = []

    for value in values:
        if value not in seen:
            seen.add(value)
            unique_items.append(value)

    return unique_items

File: Sprint-1/JavaScript/removeDuplicates/removeDuplicates.mjs
/**
 * Remove duplicate values from a sequence, preserving the order of the first occurrence of each value.
 *
 * Time Complexity:
 * Space Complexity:
 * Optimal Time Complexity:
 *
 * @param {Array} inputSequence - Sequence to remove duplicates from
 * @returns {Array} New sequence with duplicates removed
 */
export function removeDuplicates(inputSequence) {
  const seen = new Set()
  const uniqueItems = []

  for (const item of inputSequence) {
    if (!seen.has(item)) {
      seen.add(item)
      uniqueItems.push(item);
    }
  }
  return uniqueItems;
}
